server/lib/oauth2-client/oauth2.getAccessToken.js

```diff
@@ -0,0 +1,100 @@
+const { AuthorizationCode } = require('simple-oauth2');
+const logger = require('../../utils/logger');
+const { OAUTH2 } = require('../../utils/constants');
+const { Error500 } = require('../../utils/httpErrors');
+
+/**
+ * @description Get new  AccessToken with client_id and secrei_id of current oauth2 integration.
+ * @param {string} serviceId - This serviceId of current integration.
+ * @param {string} userId - This userId of current session.
+ * @param {string} authorizationCode - Authorization code for nex token.
+ * @param {string} referer - Gladys referer.
+ * @returns {Promise} Resolve with current integration service id.
+ * @example
+ * oauth2.getAccessToken(
+ *  userId: fd81vs687f0bf3414e0fe3facfba7be9455109409a'
+ *  serviceId: 'ffsdvs687f0bf3414e0fe3facfba7be945510fds09a'
+ *  authorizationCode:  '25dbf26a066d947fca82a1f05ae5890f79a27687',
+ *  referer: 'http://localhost:1444/'
+ *  }
+ * );
+ */
+async function getAccessToken(serviceId, userId, authorizationCode, referer) {
+  // Find provider configuration
+  const tokenHost = await this.variable.getValue(OAUTH2.VARIABLE.TOKEN_HOST, serviceId);
+  const tokenPath = await this.variable.getValue(OAUTH2.VARIABLE.TOKEN_PATH, serviceId);
+  const authorizeHost = await this.variable.getValue(OAUTH2.VARIABLE.AUTHORIZE_HOST, serviceId);
+  const authorizePath = await this.variable.getValue(OAUTH2.VARIABLE.AUTHORIZE_PATH, serviceId);
+  const grantType = await this.variable.getValue(OAUTH2.VARIABLE.GRANT_TYPE, serviceId);
+  const additionalAccessTokenRequestAxtionParam = await this.variable.getValue(
+    OAUTH2.VARIABLE.ADDITIONAL_ACCESS_TOKEN_REQUEST_ACTION_PARAM,
+    serviceId,
+  );
+  const redirectUriSuffix = await this.variable.getValue(OAUTH2.VARIABLE.REDIRECT_URI_SUFFIX, serviceId);
+
+  const clientId = await this.variable.getValue(OAUTH2.VARIABLE.CLIENT_ID, serviceId, userId);
+  const secret = await this.variable.getValue(OAUTH2.VARIABLE.CLIENT_SECRET, serviceId, userId);
+
+  // Init credentials based on integration name
+  const credentials = {
+    client: {
+      id: clientId,
+      secret,
+    },
+    auth: {
+      tokenHost,
+      tokenPath,
+      authorizeHost,
+      authorizePath,
+    },
+  };
+
+  // Build token access request
+  const tokenConfig = {
+    code: authorizationCode,
+    client_id: clientId,
+    client_secret: secret,
+    grant_type: grantType,
+    redirect_uri: `${referer}${redirectUriSuffix}`,
+  };
+  if (additionalAccessTokenRequestAxtionParam) {
+    tokenConfig.action = additionalAccessTokenRequestAxtionParam;
+  }
+
+  try {
+    const client = new AuthorizationCode(credentials);
+    const authResult = await client.getToken(tokenConfig, { json: true });
+
+    if (authResult.token) {
+      if (authResult.token.status && authResult.token.status !== 0) {
+        throw new Error500('Oauth2 get token response is not with status 0');
```

Comment:
why do you throw 500 errors ? 

500 errors is "The HTTP status code 500 is a generic error response. It means that the server encountered an unexpected condition that prevented it from fulfilling the request. This error is usually returned by the server when no other error code is suitable."

Here, you know what's happening, so by definition is not a 500

Reply:
I change by BadOauth2ClientResponse